File: tests/algorithms/x/Java/greedy_methods/fractional_cover_problem.java
public class Main {
    static class Item {
        int weight;
        int value;
        Item(int weight, int value) {
            this.weight = weight;
            this.value = value;
        }
        Item() {}
        @Override public String toString() {
            return String.format("{'weight': %s, 'value': %s}", String.valueOf(weight), String.valueOf(value));
        }
    }

    static Item[] items1;
    static Item[] items2;
    static Item[] items3;
    static Item[] items4;

    static double ratio(Item item) {
        return (((Number)(item.value)).doubleValue()) / (((Number)(item.weight)).doubleValue());
    }

    static double fractional_cover(Item[] items, int capacity) {
        if (capacity < 0) {
            throw new RuntimeException(String.valueOf("Capacity cannot be negative"));
        }
        double total = 0.0;
        int remaining = capacity;
        java.util.List<Item> sorted = new java.util.ArrayList<Item>() {{ java.util.ArrayList<Item> _tmp = new java.util.ArrayList<>(); for (var it : items) { _tmp.add(it); } java.util.ArrayList<Item> list = _tmp; java.util.ArrayList<Item> _res = new java.util.ArrayList<>(); list.sort((a, b) -> {Comparable _va = (Comparable)(ratio(a)); Comparable _vb = (Comparable)(ratio(b)); return _vb.compareTo(_va);}); int skip = 0; int take = -1; for (int i = 0; i < list.size(); i++) { if (i < skip) continue; if (take >= 0 && i >= skip + take) break; _res.add((Item)list.get(i)); } addAll(_res);}};
        int idx = 0;
        while (idx < String.valueOf(sorted).length() && remaining > 0) {
            Item item = sorted[idx];
            int take = item.weight < remaining ? item.weight : remaining;
            total = total + (((Number)(take)).doubleValue()) * ratio(item);
            remaining = remaining - take;
            idx = idx + 1;
        }
        return total;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            items1 = ((Item[])(new Item[]{new Item(10, 60), new Item(20, 100), new Item(30, 120)}));
            System.out.println(_p(fractional_cover(((Item[])(items1)), 50)));
            items2 = ((Item[])(new Item[]{new Item(20, 100), new Item(30, 120), new Item(10, 60)}));
            System.out.println(_p(fractional_cover(((Item[])(items2)), 25)));
            items3 = ((Item[])(new Item[]{}));
            System.out.println(_p(fractional_cover(((Item[])(items3)), 50)));
            items4 = ((Item[])(new Item[]{new Item(10, 60)}));
            System.out.println(_p(fractional_cover(((Item[])(items4)), 5)));
            System.out.println(_p(fractional_cover(((Item[])(items4)), 1)));
            System.out.println(_p(fractional_cover(((Item[])(items4)), 0)));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/PHP/greedy_methods/fractional_cover_problem.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _len($x) {
    if ($x === null) { return 0; }
    if (is_array($x)) { return count($x); }
    if (is_string($x)) { return strlen($x); }
    return strlen(strval($x));
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function ratio($item) {
  global $items1, $items2, $items3, $items4;
  return (floatval($item['value'])) / (floatval($item['weight']));
};
  function fractional_cover($items, $capacity) {
  global $items1, $items2, $items3, $items4;
  if ($capacity < 0) {
  _panic('Capacity cannot be negative');
}
  $total = 0.0;
  $remaining = $capacity;
  $sorted = [];
foreach ($items as $it) {
  $sorted[] = $it;
}
;
  $idx = 0;
  while ($idx < _len($sorted) && $remaining > 0) {
  $item = $sorted[$idx];
  $take = ($item['weight'] < $remaining ? $item['weight'] : $remaining);
  $total = $total + (floatval($take)) * ratio($item);
  $remaining = $remaining - $take;
  $idx = $idx + 1;
};
  return $total;
};
  $items1 = [['weight' => 10, 'value' => 60], ['weight' => 20, 'value' => 100], ['weight' => 30, 'value' => 120]];
  echo rtrim(_str(fractional_cover($items1, 50))), PHP_EOL;
  $items2 = [['weight' => 20, 'value' => 100], ['weight' => 30, 'value' => 120], ['weight' => 10, 'value' => 60]];
  echo rtrim(_str(fractional_cover($items2, 25))), PHP_EOL;
  $items3 = [];
  echo rtrim(_str(fractional_cover($items3, 50))), PHP_EOL;
  $items4 = [['weight' => 10, 'value' => 60]];
  echo rtrim(_str(fractional_cover($items4, 5))), PHP_EOL;
  echo rtrim(_str(fractional_cover($items4, 1))), PHP_EOL;
  echo rtrim(_str(fractional_cover($items4, 0))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
